text/0004-api-history-schema.md

```diff
@@ -0,0 +1,216 @@
+<!--
+```YAML history
+added:
+  - pr-url: https://github.com/electron/electron/pull/22533
+```
+-->
+
+* `position` [Point](structures/point.md)
+
+Set a custom position for the traffic light buttons. Can only be used with
+`titleBarStyle` set to `hidden`.
+````
+
+Which, after several PRs, may reach this state:
+
+```yaml
+added:
+  - pr-url: https://github.com/electron/electron/pull/22533
+changes:
+  - pr-url: https://github.com/electron/electron/pull/26789
+    description: Made `trafficLightPosition` option work for `customButtonOnHover` window.
+deprecated:
+  - pr-url: https://github.com/electron/electron/pull/37094
+    breaking-changes-header: deprecated-browserwindowsettrafficlightpositionposition
+removed:
+  - pr-url: https://github.com/electron/electron/pull/39479
+    breaking-changes-header: removed-browserwindowsettrafficlightpositionposition
+```
+
+Meanwhile, the Electron website displays this information nicely in a table.
+Users can then use this information to help them upgrade their Electron version,
+migrate from deprecated APIs, etc.
+
+The schema provides information to the developer about what to include in the
+block, the minimum/maximum length of text, the required properties, etc.
+
+Documentation for older APIs will also eventually be changed to include these
+API history blocks.
+
+## Reference-level explanation
+
+The JSON Schema looks like this:
+
+```json
+{
+  "title": "JSON schema for API history blocks in @electron/electron documentation",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$comment": "If you change this schema, remember to edit the TypeScript interfaces in the linting script.",
+  "definitions": {
+    "changeSchema": {
+      "type": "object",
+      "properties": {
+        "pr-url": { "type": "string", "pattern": "^https://github.com/electron/electron/pull/\\d+$" },
+        "breaking-changes-header": { "type": "string", "minLength": 3 },
+        "description": { "type": "string", "minLength": 3, "maxLength": 72 }
+      },
+      "required": [ "pr-url" ],
+      "additionalProperties": false
+    }
+  },
+  "type": "object",
+  "properties": {
+    "added": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } },
+    "deprecated": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } },
+    "removed": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } },
+    "changes": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } }
+  },
+  "additionalProperties": false
+}
+```
+
+The developer will include information about when/where/how/why an API was:
+
+- Added
+- Changed
+- Deprecated
+- Removed
+
+Along with this, they will include a link to the PR where this change was made
+and preferably include a short description of the change. If applicable, they
+will also include the header for that change from the
+[breaking changes documentation.](https://github.com/electron/electron/blob/main/docs/breaking-changes.md)
```

Comment:
If we are including breaking changes header, does it make sense to have the properties aligned with the types of breaking changes? Specifically splitting "changed" => "API changed"  , "Behavior changed", "Default changed"

Reply:
I think adding a `type` sub property to `changes` would be better e.g.:

```yaml
added:
  - pr-url: https://github.com/electron/electron/pull/1128
changes:
  - pr-url: https://github.com/electron/electron/pull/41610
  	type: behaviour
    description: `nativeImage.toDataURL` will preserve PNG colorspace
    breaking-changes-header: behavior-changed-nativeimagetodataurl-will-preserve-png-colorspace
```

Although, `electron/website` could just lookup the header in `breaking-changes.md` and display the type of change automagically. Would love to see if other people prefer it to be explicitly stated in the history block. @dsanders11 @erickzhao ?

Reply:
Pulling this out of the Summer of Code call - after reviewing the actual breaking changes doc, most of the "Changed" changes fall under "Behavior", and the differences between API/behavaior/default is fairly minor. I think we can leave this schema as "Changed" for now, and potentially add sub types in the future 🙂